### PR TITLE
NNVP-120 Bisnode API

### PR DESCRIPTION
**Pull request by jnsvd**

### CODE REVIEW

**Comment by jnsvd** [comment]
:star: PR built and available in a preview environment **the-glue-nnvp-backend-pr-67** [here](http://nnvp-backend.jx-the-glue-nnvp-backend-pr-67.cluster.development.nnvp.thegluesandbox.com)

File: src/main/java/com/theglue/nnvp/enums/BisnodeBreakType.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+@AllArgsConstructor
+@Getter
+public enum BisnodeBreakType implements NnvpEnum<BisnodeBreakType> {
+
+  STREETS,
+  CITIES;
+
```
Comment (eryk7b) [inline src/main/java/com/theglue/nnvp/enums/BisnodeBreakType.java:12]: we could go for something like

```
  STREETS(BisnodeResponseDTO::setStreets),
  CITIES(BisnodeResponseDTO::setCities);

  private BiConsumer<BisnodeResponseDTO, List<BisnodeBreakResponseDTO>> responseMapping;
```

to simplify the mapping - will add a comment there

File: src/main/java/com/theglue/nnvp/model/bisnode/Road65ResultResponse.java [hunk 2]
```diff
@@ -69,4 +69,8 @@
 
   @XmlElement(name = "house_number")
   private String houseNumber;
+
+  @XmlElement(name = "house_number_alfa")
+  private String houseNumberAlfa;
+
 }
```
Comment (eryk7b) [inline src/main/java/com/theglue/nnvp/model/bisnode/Road65ResultResponse.java:76]: we can define enums here
```
  private BisnodeLanguage cityLanguage;

  @XmlElement(name = "nis5")
  private String nis5;

  @XmlElement(name = "street_name")
  private String streetName;

  @XmlElement(name = "street_id")
  private String streetId;

  @XmlElement(name = "street_status")
  private BisnodeStreetStatus streetStatus;

  @XmlElement(name = "street_language")
  private BisnodeLanguage streetLanguage;
```

File: src/main/java/com/theglue/nnvp/controller/mapper/Road65BisnodeMappingConfig.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        .field("result.streetId", "streetId")
+        .field("result.cityId", "cityId")
+        .field("result.nis5", "nis5")
+        .field("result.nis6", "nis6")
+        .field("result.nis9", "nis9")
+        .field("result.r65extra1", "r65extra1")
+        .field("result.r65extra2", "r65extra2")
+        .field("result.houseNumber", "houseNumber")
+        .field("result.houseNumberAlfa", "houseNumberAlfa")
+        .byDefault()
```
Comment (eryk7b) [inline src/main/java/com/theglue/nnvp/controller/mapper/Road65BisnodeMappingConfig.java:39]: with enums mapped we can go 
```
        .field("result.streetStatus", "streetStatus")
        .field("result.cityLanguage", "cityLanguage")
        .field("result.streetLanguage", "streetLanguage")
```
and remove it from mapAtoB

File: src/main/java/com/theglue/nnvp/controller/mapper/Road65BisnodeMappingConfig.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+      Optional.ofNullable(source.getResult())
+          .ifPresent(result -> {
+            Optional.ofNullable(result.getStreetStatus())
+                .ifPresent(streetStatus -> target.setStreetStatus(BisnodeStreetStatus.valueOf(streetStatus)));
+            Optional.ofNullable(result.getStreetLanguage())
+                .ifPresent(streetLanguage -> target.setStreetLanguage(BisnodeLanguage.valueOf(streetLanguage)));
+            Optional.ofNullable(result.getCityLanguage())
+                .ifPresent(cityLanguage -> target.setCityLanguage(BisnodeLanguage.valueOf(cityLanguage)));
+          });
```
Comment (eryk7b) [inline src/main/java/com/theglue/nnvp/controller/mapper/Road65BisnodeMappingConfig.java:82]: can be removed if enums mapped in xml class already

File: src/main/java/com/theglue/nnvp/controller/mapper/Road65BisnodeMappingConfig.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+                  .stream()
+                  .map(this::mapBreak)
+                  .collect(Collectors.toList());
+
+              if (BisnodeBreakType.CITIES == breakType) {
+                target.setCities(elements);
+              } else {
+                target.setStreets(elements);
+              }
+            });
```
Comment (eryk7b) [inline src/main/java/com/theglue/nnvp/controller/mapper/Road65BisnodeMappingConfig.java:72]: this method with enum with biconsumer and other changes could be simplified to

```

    @Override
    public void mapAtoB(final Road65Response source, final BisnodeResponseDTO target, final MappingContext context) {
      Optional.ofNullable(source.getBreakList())
          .ifPresent(breakList -> {
            BisnodeBreakType breakType = NnvpEnum.fromCode(breakList.getType(), BisnodeBreakType.class);
            target.setBreakType(breakType);
            List<BisnodeBreakResponseDTO> elements = breakList.getLine()
                .stream()
                .map(this::mapBreak)
                .collect(Collectors.toList());
            breakType.getResponseMapping().accept(target, elements);
          });
    }
```

File: src/main/java/com/theglue/nnvp/controller/BisnodeController.java [hunk 6]
```diff
@@ -46,34 +46,63 @@ public BisnodeController(MapperProvider mapperProvider, BisnodeService bisnodeSe
...
+      BisnodeResponseDTO bisnodeResponseDto = mapperProvider.convert(road65, BisnodeResponseDTO.class);
+      bisnodeResponseDto.setRequestId(bisnodeRequest.getRequestId());
+
+      return new ResponseEntity<>(SuccessResponseDTO.successResponse(
+          bisnodeResponseDto),
+          HttpStatus.OK
+      );
+    }
+
+    if (!emptyZipCode && !emptyStreetName && allOthersAreNull) {
```
Comment (eryk7b) [inline src/main/java/com/theglue/nnvp/controller/BisnodeController.java:74]: can be just
`    if (!emptyStreetName) {
`

**Comment by jnsvd** [comment]
:star: PR built and available in a preview environment **the-glue-nnvp-backend-pr-67** [here](http://nnvp-backend.jx-the-glue-nnvp-backend-pr-67.cluster.development.nnvp.thegluesandbox.com)

**Comment by jnsvd** [comment]
:star: PR built and available in a preview environment **the-glue-nnvp-backend-pr-67** [here](http://nnvp-backend.jx-the-glue-nnvp-backend-pr-67.cluster.development.nnvp.thegluesandbox.com)

**Pull request merged**